src/main/java/net/neoforged/neoforge/common/crafting/SizedIngredient.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.stream.Stream;
+import net.minecraft.network.RegistryFriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.tags.TagKey;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.level.ItemLike;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Standard implementation for an ingredient and a count. {@link Ingredient} does not perform count checks, so this class is used to wrap an ingredient with a count, and provide a standard serialization format.
+ */
+public final class SizedIngredient {
```

Comment:
Just as a precaution, I would rename this class to something that doesn't end in just `Ingredient` because it makes it seem like this is a neoforge supported ingredient, when it is just a helper to handle sized stacks.

Reply:
You are going to run into that problem again if fluid ignredients ever get added to forge, its going to be hard to come up with an alternative that does not just make things less clear

Reply:
This class is intended to be used as a "sized ingredient" for all intents and purposes. Many mods have a hacky `Ingredient` implementation that also adds count information. This will hopefully allow them to find the more idiomatic class. I'm not against changing the name, but I think it will be hard to find something as clear.

Reply:
Best alternative I can think of is `IngredientWithSize` or `IngredientWithCount`, though it's a less recognizable name and still could be mistaken

Reply:
You could also just put the file in an inner package to separate it from the other ingredients. I just think some separation would be good if possible. If not, we can resolve this comment.